City-of-Helsinki/hel-fi-drupal-grants
# fix: AU-1809: Oma asiointi page performance issues

# [AU-1809](https://helsinkisolutionoffice.atlassian.net/browse/AU-1809)

## What was done
This pull request implements performance improvements for the "**Oma asiointi**" page. Previously, a user with 100 applications would have to wait around 9-10 seconds for the page to load. Now, the page loads in around 4 seconds with the same amount of applications.

## Things that have changed
- The `list.js` library has been updated from version  `1.3.0` to `1.5.0` in order to get pagination to work.
- The "**Sent applications**" listing on the "**Oma asiointi**" page has been paginated. It should now display 10 applications on each page.
- The logic in which we fetch a users applications on the "**Oma asiointi**" page has been modified. Previously, multiple unnecessary calls to load an applications data were being made during the request. Now, the data is fetched once and passed on to the parts of the code that actually need it.
- The `ApplicationListingController` class has been deleted, since it wasn't being used by any code.

## Testing results
I have been testing the new implementation with Jannes test user, since it has a lot of applications. If you want to do the same, here are the credentials:

In `local.settings.php`:
- `putenv('APP_ENV=localj');`

Login credentials:
- `100384-998R` -> select the registered community role.

### The testing results I've achieved are the following:

Loading the `/fi/oma-asiointi` page without the new changes
1. 9.50 s
2. 9.72 s
3. 8.99 s
4. 9.62 s
5. 10.06 s
6. 10.91 s
7. 9.31 s
8. 9.71 s
9. 8.91 s

Loading the `/fi/oma-asiointi` page with the new changes
1. 4.55 s
2. 3.84 s
3. 3.85 s
4. 4.42 s
5. 3.92 s
6. 4.21 s
7. 4.05 s
8. 3.69 s
9. 3.85 s

## How to install
Make sure your instance is up and running on the correct branch.
  * `git checkout feature/AU-1809-oma-asiointi-pagination`
  * `make fresh`
  * `make drush-cr`

## How to test
Before you begin, set this variable in your `local.settings.php` file: `putenv('ATV_PAGE_SIZE=100');` and run `drush cr`. This changes the size of the request made to ATV, meaning that if we have many applications, we have to make fewer requests. This change also needs to be implemented in [Azure](https://dev.azure.com/City-of-Helsinki/avustusasiointi/_library?itemType=VariableGroups) once this pull request is merged into production.

- Start by benchmarking your load speeds on the development branch. Login with a user that has around 100 applications (use Jannes user mentioned above if needed). Go to the [Oma asiointi ](https://hel-fi-drupal-grant-applications.docker.so/fi/oma-asiointi) page and perform a hard refresh (cmd+shift+r on a mac). Refresh the page a few times and check how long it takes to load. Loading speeds can be found under the network tab in the browsers developer console.

- Now, do the same tests in this branch. The load speeds should be around 60% faster compared to the development branch.

- Test that paginating, filtering and sorting the applications works.

- Review the code.

[AU-1809]: https://helsinkisolutionoffice.atlassian.net/browse/AU-1809?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
